convention.md

```diff
@@ -324,3 +332,28 @@ For example, an organization `example.org` wanting to add a feature `our-feature
 Every extension must be published using a license.
 The license can be chosen freely, even proprietary licenses are possible.
 The recommended license is the [CCA 4.0](https://homieiot.github.io/license), since this is the license Homie itself uses.
+
+## Implementation notes
+
+### Device initialization
+
+Some devices require knowledge of their settable retained properties to function properly.
+The homie convention does not specify how to initialize/recover them e.g. after a power cycle.
+A few common approaches are:
+
+* A device can simply load default values from some configuration file.
+* A device can restore its previous state from some local storage. This is the recommended way.
+* A device may try to restore its state using MQTT. This can be done by subcribing to the respective channels.
+An alternative way is to recover the state from other MQTT channels that are external to the Homie specification.
+This is not a recommended approach though, because retained messages are only sent by the broker in response to a new subscription.
+So if a device doesn't reconnect with a clean session, then the retained messages won't be resent.
```

Comment:
If a device connects with a non clean session it is assumed to "know" what happend in the last connection (mqtt kind of defines this). This means it should still know the retained values sent. If there is no persistence it should really be connecting with clean session set.

Reply:
I'm sorry, but I don't understand. What exactly is a "clean session"? What exactly is wrong about the paragraph and what should a more correct version?

My motivation behind this was to say that this approach at restoring state is a bad idea, and to explain why.

Reply:
MQTT supports keeping sessions while the device is offline.
If you connect and request a clean session you will get one that looks as if you never connected before.
However if you dont request a clean session the broker will restore all subscriptions you had the last you where connected.  It will also deliver all messages sent to a subscription with qos 1 or higher while the device was offline.

You would normally connect with a clean session if your device state is lost (like after a reboot without persistent storage, or after a reset) and connect to an existing session otherwise if you're device still knows what state its world is in (like a connection drop/reconnect).

The only difference is that with an existing session you dont get retained topics unless you unsubscribe and resubscribe them again or they changed while the device was offline. Also there is no need to subscribe if you have an existing session as the broker keeps subscriptions alive.

Reply:
@homieiot/involved in my implementations, the Homie devices know their state ie on/off, dimmer percentage, temperature etc. Is there a situation where a device would not know it's own state and need to restore a value?

As part of other discussion regarding QoS, the Homie spec probably needs to address controller set messages that are not delivered in a timely fashion.

Reply:
@mjcumming The section I wrote about this is limited to settable properties that can't easily be recovered. Think of music players for example. Or in my case, a light effects controller.